File: C/InsertionSort.c
// simple example of an insertion sort on an array

#include <stdio.h>

// Function Description:
// - performs insertion sort on an array
// - builds the sorted array one item at a time by inserting elements
// - Note: I prefer not having to use a break from approach in the while loop
// Arguments:
// - unsortedArray - integer array to sort
// - unsortedSize - size_t length of the array
// Return Value:
// - array passed by reference
static void insertionSort(int *unsortedArray, size_t unsortedSize)
{
	// outer loop controlling the number of sort passes
	for (int arrayItem = 1; arrayItem < unsortedSize; ++arrayItem)
	{
		// get current element
		int currentItem = unsortedArray[arrayItem];
		// the position of the division of unsorted vs sorted
		int insertItem = arrayItem - 1;

		// while value of insertItem is > currentItem
		while ((insertItem >= 0 && unsortedArray[insertItem] > currentItem))
		{
			// move insertItem to the right
			unsortedArray[insertItem + 1] = unsortedArray[insertItem];
			// move backward through the list to check the next insertion
			--insertItem;
		}
		// insert currentItem to the right
		unsortedArray[insertItem + 1] = currentItem;
	} // for each arrayItem
} // insertion sort

// main entry point
int main() {
	int arrayToSort[20];
	int arraySize = sizeof(arrayToSort) / sizeof(arrayToSort[0]);
	srand(time(NULL));

	// populate array with random values
	for (int index = 0; index < arraySize; index++)
	{
		arrayToSort[index] = rand() % 100;
	} // end of for	

	printf("Array prior to sort: \n");
	for (int index = 0; index < arraySize; index++)
	{
		printf("%d%s", arrayToSort[index], " ");
	}

	// perform insertion sort
	insertionSort(&arrayToSort, (sizeof(arrayToSort) / sizeof(*arrayToSort)));

	printf("\n\nArray after sort: \n");
	for (int index = 0; index < arraySize; index++)
	{
		printf("%d%s", arrayToSort[index], " ");
	}

	printf("\n");
} // main

File: Python/InsertionSort.py
# simple example of an insertion sort on an array
import numpy

# Function Description:
# - performs insertion sort on an array
# - builds the sorted array one item at a time by inserting elements
# Arguments:
# - unsortedArray - integer array to sort
# Return Value:
# - array passed by reference
def insertion_sort(unsorted_array):
    # for each item in the array
    for outer_pass in range(1, len(unsorted_array)):
        # get current element
        current_element = unsorted_array[outer_pass]
        # the position of the division of unsorted vs sorted
        insertion = outer_pass - 1

        # while value of insertion is > current_element
        while (insertion >= 0) and (current_element < unsorted_array[insertion]):
            # move insertion to the right
            unsorted_array[insertion + 1] = unsorted_array[insertion]
            # move backward through the list to check the next insertion
            insertion -= 1

        # insert currentItem to the right
        unsorted_array[insertion + 1] = current_element

# test function / main entry point

# populate array with random values
array_to_sort = numpy.random.randint(0, 100, size=20)

print("Array prior to sort: ")
print(array_to_sort)

insertion_sort(array_to_sort)

print("Array after sort: ")
print(array_to_sort)
